telethon/network/connection/connection.py

```diff
@@ -46,58 +59,76 @@ def __init__(self, ip, port, dc_id, *, loggers, proxy=None, local_addr=None):
             else:
-                local_addr = None
+                raise ValueError("Unknown local address format: {}".format(self._local_addr))
+        else:
+            local_addr = None
 
-            self._reader, self._writer = await asyncio.wait_for(
-                asyncio.open_connection(self._ip, self._port, ssl=ssl, local_addr=local_addr),
-                timeout=timeout
-            )
+        if not self._proxy:
+            connect_coroutine = asyncio.open_connection(
+                host=self._ip,
+                port=self._port,
+                ssl=ssl,
+                local_addr=local_addr)
         else:
-            import socks
-            if ':' in self._ip:
-                mode, address = socket.AF_INET6, (self._ip, self._port, 0, 0)
-            else:
-                mode, address = socket.AF_INET, (self._ip, self._port)
 
-            s = socks.socksocket(mode, socket.SOCK_STREAM)
-            if isinstance(self._proxy, dict):
-                s.set_proxy(**self._proxy)
+            if isinstance(self._proxy, (tuple, list)):
+                proxy, auth, remote_resolve = self._parse_proxy(*self._proxy)
+            elif isinstance(self._proxy, dict):
+                proxy, auth, remote_resolve = self._parse_proxy(**self._proxy)
             else:
-                s.set_proxy(*self._proxy)
-
-            s.settimeout(timeout)
-            if self._local_addr is not None:
-                s.bind((self._local_addr, None))
-            await asyncio.wait_for(
-                asyncio.get_event_loop().sock_connect(s, address),
-                timeout=timeout
-            )
-            if ssl:
-                if ssl_mod is None:
-                    raise RuntimeError(
-                        'Cannot use proxy that requires SSL'
-                        'without the SSL module being available'
-                    )
-
-                s = ssl_mod.wrap_socket(
-                    s,
-                    do_handshake_on_connect=True,
-                    ssl_version=ssl_mod.PROTOCOL_SSLv23,
-                    ciphers='ADH-AES256-SHA'
-                )
-                
-            s.setblocking(False)
-
-            self._reader, self._writer = await asyncio.open_connection(sock=s)
+                raise ValueError("Unknown proxy format: {}".format(self._proxy.__class__.__name__))
 
+            connect_coroutine = aiosocks.open_connection(
+                proxy=proxy,
+                proxy_auth=auth,
+                dst=(self._ip, self._port),
+                remote_resolve=remote_resolve,
+                ssl=ssl,
+                local_addr=local_addr)
+
+        self._reader, self._writer = await asyncio.wait_for(connect_coroutine, timeout=timeout)
         self._codec = self.packet_codec(self)
         self._init_conn()
         await self._writer.drain()
 
+    @staticmethod
+    def _parse_proxy(protocol, host, port, username=None, password=None, remote_resolve=True):
```

Comment:
This is a breaking change. We *can* use `aiosocks` underneath, but the public interface must continue to support the old proxy style. So we need to be able to translate `pysocks` usage into `aiosocks` usage (old users will have two proxy libraries unless they update to the new style, but that's not a problem).

Meaning users will be able to use this:

```python
(socks.SOCKS5, host, port)
```

And Telethon will translate that into the following:

```python
aiosocks.Socks5Addr(host, port)
```

And all the other tuple lengths too, as indicated in PySocks README.

Reply:
@Lonami 

Commited the change to signature of `_parse_proxy()`.
Now it should be compatible with both `tuple` and `dict` formats of `PySocks`.

Note that `socks.SOCKS5 == 2` and `socks.SOCSK4 == 1` so that we do not need to actually import
`PySocks` to do the checks, but rather just check against integer values!